docs/map.rst

```diff
@@ -0,0 +1,602 @@
+``Data.Map.Strict``, or if your keys are ``Int`` use ``Data.IntMap``; all of
+these implementations are *immutable*.
+
+::
+
+    data Map k v = ...
+
+    data IntMap v = ...
+
+.. IMPORTANT::
+   ``Map`` relies on the key type ``k`` having instances of the ``Eq`` and
+   ``Ord`` typeclass for its internal representation. These are already defined
+   for builtin types, and if you are using your own data type you can use the
+   `deriving
+   <https://en.wikibooks.org/wiki/Haskell/Classes_and_types#Deriving>`_
+   mechanism.
+
+
+Short Example
+-------------
+
+The following GHCi session shows some of the basic map functionality::
+
+    import qualified Data.Map.Strict as Map
+
+    let nums = Map.fromList [(1,"one"), (2,"two"), (3,"three")]
+
+    -- Check if we know the english words for 1 and 4
+    Map.member nums 1
+    > True
+
+    Map.member nums 4
+    > False
+
+
+    -- Get the english word for the number 3 and 4.
+    Map.lookup 3 nums
+    > Just "three"
+
+    Map.lookup 4 nums
+    > Nothing
+
+
+    -- Add (4, "four") to our original map.
+    let moreNums = Map.insert 4 "four" nums
+
+    Map.member moreNums 4
+    > True
+
+
+    -- Remove the entry for 1 from our original map.
+    let fewerNums = Map.delete 1 nums
+
+    Map.toAscList fewerNums
+    > [(2,"two"),(3,"three")]
+
+
+    -- Create a new map and combine it with our original map.
+    -- Note that these operations are left-biased.
+    let newNums = Map.fromList [(3,"new three"), (4,"new four")]
+
+    Map.union newNums nums
+    > fromList [(1,"one"),(2,"two"),(3,"new three"),(4,"new four")]
+
+
+.. TIP:: You can use the `OverloadedLists
+	 <https://ghc.haskell.org/trac/ghc/wiki/OverloadedLists>`_ extension so
+	 you don't need to write ``fromList [1, 2, 3]`` everywhere; instead you
+	 can just write ``[1, 2, 3]`` and if the function is expecting a map it
+	 will be converted automatically! The code here will continue to use
+	 ``fromList`` for clarity though.
+
+
+Importing Map and IntMap
+------------------------
+
+When using ``Map`` or ``IntMap`` in a Haskell source file you should always use
+a ``qualified`` import because these modules export names that clash with the
+standard Prelude (you can import the type constructor on its own though!). You
+should also import Prelude and hide ``lookup`` because if you accidentally leave
+off the ``Map.`` qualifier you'll get confusing type errors.
+
+::
+
+    import Prelude hiding (lookup)
+
+    import Data.Map.Strict (Map)
+    import qualified Data.Map.Strict as Map
+
+    import Data.IntMap (IntMap)
+    import qualified Data.IntMap.Strict as IntMap
+
+
+Common API Functions
+--------------------
+
+.. TIP::
+   All of these functions that work for ``Map`` will also work for ``IntMap``,
+   which has the key type ``k`` specialized to ``Int``. Anywhere that you
+   see ``Map Int v`` you can replace it with ``IntMap v``.
+
+.. NOTE::
+   ``fromList [some,map,entries]`` is how a ``Map`` is printed.
```

Comment:
I would use an example demonstrating that the list is made of *pairs*. Maybe

> A `Map` is printed as an association list preceded by `fromList`. For example, it might look like
> ```haskell
> fromList [(Key1, True), (Key2, False)]
> ```

Reply:
Done.